app/src/main/java/com/lagradost/cloudstream3/ui/result/ResultViewModel2.kt

```diff
@@ -841,71 +882,188 @@ class ResultViewModel2 : ViewModel() {
 
         if (isFavorite) {
             removeFavoritesData(currentId)
+            statusChangedCallback?.invoke(false)
+            _favoriteStatus.postValue(false)
+            return false
         } else {
-            val current = getFavoritesData(currentId)
+            checkAndWarnDuplicates(
+                context,
+                R.string.duplicate_message_favorites,
+                response.name,
+                response.year,
+                response.getImdbId(),
+                getAllFavorites(),
+            ) { shouldContinue: Boolean, duplicateId: Int? ->
+                if (!shouldContinue) {
+                    statusChangedCallback?.invoke(null)
+                    return@checkAndWarnDuplicates
+                }
 
-            setFavoritesData(
-                currentId,
-                DataStoreHelper.FavoritesData(
+                if (duplicateId != null) {
+                    removeFavoritesData(duplicateId)
+                }
+
+                val current = getFavoritesData(currentId)
+
+                setFavoritesData(
                     currentId,
-                    current?.favoritesTime ?: unixTimeMS,
-                    unixTimeMS,
-                    response.name,
-                    response.url,
-                    response.apiName,
-                    response.type,
-                    response.posterUrl,
-                    response.year
+                    DataStoreHelper.FavoritesData(
+                        current?.favoritesTime ?: unixTimeMS,
+                        currentId,
+                        unixTimeMS,
+                        response.name,
+                        response.url,
+                        response.apiName,
+                        response.type,
+                        response.posterUrl,
+                        response.year,
+                        response.getImdbId()
+                    )
                 )
-            )
+
+                _favoriteStatus.postValue(true)
+
+                statusChangedCallback?.invoke(true)
+            }
+
+            return _favoriteStatus.value
+        }
+    }
+
+    private fun checkAndWarnDuplicates(
+        context: Context?,
+        message: Int,
+        name: String,
+        year: Int?,
+        imdbId: String?,
+        data: List<DataStoreHelper.LibrarySearchResponse>,
+        checkDuplicatesCallback: (shouldContinue: Boolean, duplicateId: Int?) -> Unit
+    ) {
+        fun normalizeString(input: String): String {
+            /**
+             * Trim the input string and replace consecutive spaces with a single space.
+             * This covers some edge-cases where the title does not match exactly across providers,
+             * and one provider has the title with an extra whitespace. This is minor enough that
+             * it should still match in this case.
+             */
+            return input.trim().replace("\\s+".toRegex(), " ")
```

Comment:
this recreates the regex for every single run of the function, and that can be really expensive if you want to check multiple items. in this case it is about 30x faster to cache the regex in a persistent variable